### PR TITLE
Feat: Add Docker file

### PR DESCRIPTION
**Pull request by pantharshit007**
Added docker file for both `backend` and `frontend`

and `docker-compose.yml` for running the whole project in one go

`docker-compose up -d`
image

`docker-compose logs backend`
image

`docker-compose logs frontend -f`
image

image

image

_I am working on the cli feature after that it is up to you to decide whether to make it optional or permanent._

### CODE REVIEW

**Comment by manishraj27** [comment]
The branch has merge conflicts please resolve those also I've deleted templates folder because it does not solve the purpose.

I appreciate your contribution!

**Comment by pantharshit007** [comment]
image
provide me write access or try making an assignee may be that solves the issue

**Comment by manishraj27** [comment]
Yeah, you have got merge issues because your fork was not up-to-date with the upstream branch!

**Comment by pantharshit007** [comment]
@manishraj27  check now

**Comment by pantharshit007** [comment]
> Yeah, you have got merge issues because your fork was not up-to-date with the upstream branch!

what I mean here is that resolve button is disabled

**Review by pantharshit007 — COMMENTED** [review]
you can check for any errors, and now you can also remove/comment the part in `createMERNProject.js`

**Comment by pantharshit007** [comment]
@manishraj27 this repository is not participating add the tag in repo description for more info check there site

**Comment by manishraj27** [comment]
you have made changes in other files as well and you haven't run 
npm run lint
npm run lint:fix 
the other command files should be untouched since you are creating a new file so there should be no dependent changes

the only files should be:
1. Readme.md
2. new file (initializeDocker.js)
3. index.js (include the initializeDocker.js)
4. templates folder (no need to create it just shows the preview so remove template folder)

**Comment by pantharshit007** [comment]
> I suggest you to take backup of your file and update with upstream and create one file initializeDocker.js and add the command to index.js.

oh I am sorry I totally forgot about updating the index.js part 
so you mean to say I should revert all the changes except in these files?
```
Readme.md
new file (initializeDocker.js)
index.js (include the initializeDocker.js)
templates folder (no need to create it just shows the preview so remove template folder)
```

so, I will be reverting changes from `CREATEmernproject.js` and `mongoConnect`

**Comment by pantharshit007** [comment]
@manishraj27 could you check now.

File: README.md [hunk 1]
```diff
@@ -567,6 +617,7 @@ Skip the boring setup and jump straight into building your next big idea! Whethe
 
 1. **Code Generation**
    More Code-Snippets
+   More Code-Snippets
```
Comment (pantharshit007) [inline README.md:620]: delete this before merge

**Comment by manishraj27** [comment]
Is template folder necessary?

**Comment by pantharshit007** [comment]
> Is template folder necessary? @pantharshit007

i don't know you guys put it there I thought its for reference

**Comment by manishraj27** [comment]
> i don't know you guys put it there I thought its for reference

Yeah, that's why I removed the template folder.
Upstream your branch and check the issue.
Also I'am trying to run the command so ill need to install docker right?

**Comment by pantharshit007** [comment]
> > i don't know you guys put it there I thought its for reference
> 
> Yeah, that's why I removed the template folder. Upstream your branch and check the issue. Also I'am trying to run the command so ill need to install docker right?

yeah you need docker for that or if you are not able to I can send a video of the process here but don't know if it will work since these command should be working based on the code uploaded on npm registery

**Comment by manishraj27** [comment]
> yeah you need docker for that or if you are not able to I can send a video of the process here but don't know if it will work since these command should be working based on the code uploaded on npm registery


ok ill install the docker but the problem is even if i am running in backend or frontend or in the project it is still telling to run in root

**Comment by pantharshit007** [comment]
> > yeah you need docker for that or if you are not able to I can send a video of the process here but don't know if it will work since these command should be working based on the code uploaded on npm registery
> 
> ok ill install the docker but the problem is even if i am running in backend or frontend or in the project it is still telling to run in root

I have added a comment above check it

**Comment by manishraj27** [comment]
> I have added a comment above check it
image
image
testDock is project which has backend and frontend folder ive run the command but check what it says

**Comment by pantharshit007** [comment]
> > I have added a comment above check it
> > image
> > image
> > testDock is project which has backend and frontend folder ive run the command but check what it says

On line 18 you still have those ! Remove them
```
If (isBackend || isFrontend)
```

**Comment by manishraj27** [comment]
> If (isBackend || isFrontend)
ok i see

**Review by manishraj27 — COMMENTED** [review]
run npm run lint and npm run lint:fix

**Comment by manishraj27** [comment]
Your initializeDocker.js code is perfect but you have made some changes in mongodbConnect as well which is creating db name statically before that it used to create db name dynamically so i want you to update the upstreat and raise the pr again with just initializeDocker.js and index.js and than run npm run lint and npm run lint:fix and then it can be merged.


the below code is the initializeDocker.js {i have made some changes in it}



import fs from 'fs-extra';
import path from 'path';
import chalk from 'chalk';

export default function initializeDockerCommand(program) {
  program
    .command('create-dockerfiles')
    .description(
      'Generate Dockerfiles for backend and frontend, and docker-compose.yml in root'
    )
    .action(() => {
      const currentDir = process.cwd();

      // Check if backend and frontend directories exist
      const backendDir = path.join(currentDir, 'backend');
      const frontendDir = path.join(currentDir, 'frontend');

      const hasBackendDir = fs.existsSync(backendDir);
      const hasFrontendDir = fs.existsSync(frontendDir);

      if (!hasBackendDir || !hasFrontendDir) {
        console.error(chalk.red('❌ Required directory structure not found.'));
        console.log(
          chalk.yellow(
            '📁 Please ensure both "backend" and "frontend" directories exist in the current directory.'
          )
        );
        return;
      }
      // Dockerfile for backend
      const backendDockerfileContent = `# Use the official Node.js image
FROM node:20-alpine

WORKDIR /app

COPY package*.json ./

RUN npm install

COPY . .

EXPOSE 5000

CMD ["npm", "run", "dev"]
`;
      const dockerIgnore = `node_modules
dist`;

      // Dockerfile for frontend
      const frontendDockerfileContent = `# Use the official Node.js image
FROM node:20-alpine

# Set the working directory
WORKDIR /app

# Copy package.json and package-lock.json
COPY package*.json ./

# Install dependencies
RUN npm install

# Copy the rest of the application files
COPY . .

# Expose the port
EXPOSE 3000

# Start the application in development mode
CMD ["npm", "start"]

`;

      // docker-compose.yml content
      const dockerComposeContent = `# version: "3.8"

services:
  backend:
    build: ./backend
    ports:
      - "5000:5000"
    volumes:
      - ./backend:/app
    environment:
      DB_URL: mongodb://mongo:27017/mydatabase # you can change db name here to your liking
      # this doesnt work since for that we need .env on same level
      # DB_URL: mongodb://mongo:27017/\${DB_NAME}

  frontend:
    build: ./frontend
    ports:
      - "3000:3000"
    volumes:
      - ./frontend:/app
    environment:
      REACT_APP_API_URL: http://localhost:5000/api
    depends_on:
      - backend # Ensure the backend is started before the frontend

  mongo:
    image: mongo:latest
    ports:
      - "27017:27017"
    volumes:
      - mongo_data:/data/db

volumes:
  mongo_data:

`;

      // Create backend Dockerfile & .dockerignore
      const backendDockerfilePath = path.join(backendDir, 'Dockerfile');
      const backendDockerIgorefilePath = path.join(backendDir, '.dockerignore');
      try {
        fs.writeFileSync(backendDockerfilePath, backendDockerfileContent);
        fs.writeFileSync(backendDockerIgorefilePath, dockerIgnore);
        console.log(chalk.green('✅ Backend Dockerfile created successfully.'));
      } catch (error) {
        console.error(
          chalk.red(`❌ Failed to create Backend Dockerfile: ${error.message}`)
        );
        return;
      }

      // Create frontend Dockerfile & .dockerignore
      const frontendDockerfilePath = path.join(frontendDir, 'Dockerfile');
      const frontendDockerIgorefilePath = path.join(
        frontendDir,
        '.dockerignore'
      );
      try {
        fs.writeFileSync(frontendDockerfilePath, frontendDockerfileContent);
        fs.writeFileSync(frontendDockerIgorefilePath, dockerIgnore);
        console.log(
          chalk.green('✅ Frontend Dockerfile created successfully.')
        );
      } catch (error) {
        console.error(
          chalk.red(`❌ Failed to create Frontend Dockerfile: ${error.message}`)
        );
        return;
      }

      // Create docker-compose.yml in root
      const dockerComposePath = path.join(currentDir, 'docker-compose.yml');
      try {
        fs.writeFileSync(dockerComposePath, dockerComposeContent);
        console.log(chalk.green('✅ docker-compose.yml created successfully.'));
      } catch (error) {
        console.error(
          chalk.red(`❌ Failed to create docker-compose.yml: ${error.message}`)
        );
      }

      // Final success message
      console.log(chalk.cyan('\n🎉 Docker files created successfully!'));
    });
}

**Comment by pantharshit007** [comment]
> run npm run lint and npm run lint:fix

Dont need since I directly merge your changes in this branch this way your commit will also be preserved.

**Review by manishraj27 — APPROVED** [review]
merged

**Comment by manishraj27** [comment]
https://github.com/manishraj27/mern-project-cli/commit/4b444f1a479217c58380787384970eec7448212e

@pantharshit007

**Pull request merged**